### DISCUSSION TITLE
Are there any documentation or samples for using WinUI3 in a MFC control/Window in an existing MFC application...?

### DISCUSSION THREAD

**Discussion by abuitech** [microsoft/WindowsAppSDK#2233] [Q&A]
Hello,
Is it possible to use WinUI3 controls in a MFC control or window in an existing MFC application?
If yes, where I can I find documentations or samples about this ?
I only found doc and sample to use WinRT controls in a MFC control using Xaml Islands.
Thank you.

**Comment by ghost** [comment]
Using WinUI 3 in any **Non-WinUI 3 App** (i.e WinUI2, MFC. WinForm. WPF, raw Win32 , custom controls etc) will require "**WinUI 3 Xaml Island**" feature. 

> ** - WinUI 3 in 1.0 and 1.1 is only supported for use by full WinUI 3 apps. You'll be able to use WinUI 3 in WPF/WinForms/other apps in a future release via XAML Islands.

currently there is no "**WinUI 3 Xaml Island**" support. it's on their [roadmap](https://portal.productboard.com/winappsdk/1-windows-app-sdk/c/9-xaml-islands-use-winui-3-from-wpf-or-winforms).

 there is  "WinUI 2 Xaml Island"  available, which is what Windows Terminal uses.

**Reply by abuitech** [reply]
Thank you very much :)

**Reply by amtopel** [reply]
But how does Terminal use the WinUI 3 tab control, if XAML Islands only hosts stuff from the Windows namespace?

**Reply by DarranRowe** [reply]
It doesn't, Terminal uses TerminalApp.TitleBarControl. You can see this on [line 350 of the non client window source file](https://github.com/microsoft/terminal/blob/main/src/cascadia/WindowsTerminal/NonClientIslandWindow.cpp#L350).
According to the [IDL file](https://github.com/microsoft/terminal/blob/main/src/cascadia/TerminalApp/TitlebarControl.idl), TitleBarControl derives from Windows.UI.Xaml.Controls.Grid, so so the regular Windows Xaml Islands is perfectly fine here.
Also, just to be clear, Windows Xaml Islands can host things from other namespaces, it is able to host WinUI 2 controls which exist in the Microsoft namespace. But they have to be derived from Windows.UI.Xaml.Controls.Control.

**Comment by castorix** [comment]
It can be done with _MddBootstrapInitialize_ but it is complicated
I just did a test in a C++/Win32 app by adding a Page + Button (on the left) : 

![Win32_WinUI3](https://user-images.githubusercontent.com/22345506/157936735-439ef8d8-3505-4864-ab38-be8c48713478.jpg)

**Reply by abuitech** [reply]
Very interesting.
Maybe, could I have this Test codes?
Thank you.

**Reply by abuitech** [reply]
I tried MddBootstrapInitialize() following instructions from this page : https://docs.microsoft.com/en-us/windows/apps/windows-app-sdk/tutorial-unpackaged-deployment?tabs=cpp

And when I instanciated a controls (_winu3iButton = winrt::Microsoft::UI::Xaml::Controls::Button();), it thrown an exception winrt::hresult_wrong_thread.

May I have your Test project source codes, please?

Thank you.

**Reply by castorix** [reply]
> And when I instanciated a controls (_winu3iButton = winrt::Microsoft::UI::Xaml::Controls::Button();), it thrown an exception winrt::hresult_wrong_thread.

You must call _Microsoft::UI::Xaml::Application::Start_, like :

            winrt::init_apartment(winrt::apartment_type::single_threaded);	
	        winrt::Microsoft::UI::Xaml::Application::Start(
		[](auto&&)
		{
			// code like
			// ::winrt::make<::winrt::namespace_name::implementation::App>();
		});

(my code is too "dirty" atm, as I copied quickly the  generated files from the Templates , like App.xaml.g.h, MainWindow.xaml.g.h,...

**Reply by abuitech** [reply]
Nice!  I think I see what to do. Thank you very much.

**Comment by Dramcryx** [comment]
XAML Islands are supported since 1.4 release. They even work, I have just managed to view some. To run them in MFC environment you cannot use `winrt::Microsoft::UI::Xaml::Application::Start` because it starts message loop as well.
You need to have these three guys initialized before you start message loop in MFC:
```
static winrt::Microsoft::UI::Dispatching::DispatcherQueueController dispatcherQueueController{ nullptr };
static winrt::Microsoft::UI::Xaml::Application xamlApp{ nullptr };
static winrt::Microsoft::UI::Xaml::Hosting::WindowsXamlManager xamlManager{ nullptr };

... before Run()
	dispatcherQueueController = winrt::Microsoft::UI::Dispatching::DispatcherQueueController::CreateOnCurrentThread();
	xamlApp = make<AppL>(); // if you plan custom controls
	xamlManager = winrt::Microsoft::UI::Xaml::Hosting::WindowsXamlManager::InitializeForCurrentThread();
... after Run()
	dispatcherQueueController.ShutdownQueue();
	xamlApp.Exit();
	xamlManager.Close(); // AFAIK not necessary
```
Making islands is the same as WinUI 2 except some minor changes in DesktopWindowXamlSource
https://github.com/microsoft/WindowsAppSDK-Samples/blob/main/Samples/Islands/cpp-win32-unpackaged/SimpleIslandApp.cpp
To have custom controls (implement `AppL`), you can refer to
https://www.codeproject.com/Articles/5359576/WinUI3-in-Existing-Win32-Applications
This works well, creates custom controls normally.

**Reply by DarranRowe** [reply]
There are two more things that the Islands sample does mention, but it is also worth mentioning here.
First, the Islands have a function to check if a message should be handled by the island itself. It is that ContentPreTranslateMessage function. It can feel as if you don't need to use this since the islands look like they work quite well without it. But you really should put it in the message pump, and if it returns true then do not dispatch the message.
Second, if the island is in a window along with other Windows API controls, then you should have functionality to navigate to and from the island. This means you need to hook up the TakeFocusRequested event to the DesktopWindowXamlSource to navigate out of the island, and you have to monitor the WM_KEYDOWN message for tab to see if you are navigating into the island.

**Reply by Dramcryx** [reply]
The first point is clear - MFC's app impl can override message loop to implement pre-translation of ContentPreTranslateMessage.
However, I do not get the point regarding WM_KEYDOWN. Stepping out of control is covered in the example code:
https://github.com/microsoft/WindowsAppSDK-Samples/blob/619f056317d5fa5b954cb4620808b0e8284c6835/Samples/Islands/cpp-win32-unpackaged/SimpleIslandApp.cpp#L244
Then what should I do if I step in the island?

**Reply by DarranRowe** [reply]
This again is covered in the sample in the [ProcessMessageForTabNavigation](https://github.com/microsoft/WindowsAppSDK-Samples/blob/main/Samples/Islands/cpp-win32-unpackaged/SimpleIslandApp.cpp#L114) function. Basically, if you are pressing tab and the next window is the island, or if you are holding shift and pressing tab and the previous window is the island, then it will use [DesktopWindowXamlSource.NavigateFocus](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.xaml.hosting.desktopwindowxamlsource.navigatefocus?view=windows-app-sdk-1.4#microsoft-ui-xaml-hosting-desktopwindowxamlsource-navigatefocus(microsoft-ui-xaml-hosting-xamlsourcefocusnavigationrequest)) to navigate into the island. The request, which is the parameter, will set the focus on the first or the last control depending on the direction you are tabbing.
This is pretty useful to point out because the sample places the call to this in the message pump, and again it doesn't dispatch the message if the function indicates that it handled the message.

**Reply by IMFJS** [reply]
We are struggling to add the Windows App SDK for Xaml Island (WinUI 3) support to our existing unpackaged MFC application. Our Solution has the added complexity that we need these dependencies only in one of our DLLs that is then included in our main project. However, I even run into problems when trying to add Xaml Islands to a new default MFC application as generated by the wizard without changing any settings.

Visual Studio Version: Microsoft Visual Studio Professional 2022 Version 17.8.8
Microsoft.WindowsAppSDK: 1.5.240311000
Microsoft.Windows.SDK.BuildTools (installed by the Windows App SDK): 10.0.22621.756
Microsoft.Windows.CppWinRT: 2.0.240111.5

First, I get a warning from MSBuild:
> ...\Microsoft Visual Studio\2022\Professional\MSBuild\Microsoft\VC\v170\Microsoft.CppBuild.targets(519,5): warning MSB8021: The value 'Dynamic' of the variable 'UseOfMFC' is incompatible with the value 'true' of the variable 'WindowsAppContainer'.

We do need MFC to be linked dynamically, so changing that is not an option (besides, it is the default). What's more is that there is no "WindowsAppContainer" variable in the vcxproj file. Adding this variable to the globals PropertyGroup and setting it to false (we do not want any kind of packaging, since we are an unpackaged application) does not change anything.

Second, there is a compiler error just after the MSBuild warning:
> ...\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.38.33130\atlmfc\include\afx.h(24,1): error C1189: #error:  Building MFC application with /MD[d] (CRT dll version) requires MFC shared dll version. Please #define _AFXDLL or do not use /MD[d] (compiling source file 'pch.cpp')

Again, we need MFC with dynamic linking (which does not sound like it is set here anymore) and also want the CRT linked dynamically.

It is really difficult to piece together all the scattered information about this, especially with the tangled mess of UWP/WinUI2 and WindowsAppSDK/WinUI3 resources online. Most resources also assume either a packaged application or a plain Win32 application.
It would be really nice if Microsoft could provide a sample project that shows how to integrate WinUI 3 through Xaml Islands in the standard MFC template project. In terms of unpackaged native C++ applications WinUI 3 is after all the first modern UI Framework since MFC, so this should be a quite common scenario.

Any help is appreciated.

**Reply by DarranRowe** [reply]
@Optimierungswerfer 

It would be nice to know what you steps you are taking in setting up your MFC project. However, you do not need to set MFC to static or use application packaging to use Xaml Islands.
In general, setting up Xaml Islands in a MFC application shouldn't be too different from setting it up in a regular desktop application. The biggest issue would be that you are likely going to hit conflicts in the headers. For example, MFC uses macros for exception handling, but one of these TRY, is used in the WinRT headers for the Turkish currency.

**Reply by IMFJS** [reply]
Thank you for your quick response.

I created a default MFC application with all settings left untouched and then tried to follow these steps: https://github.com/microsoft/WindowsAppSDK-Samples/tree/main/Samples/Islands#how-to-add-an-island-with-xaml-content-to-your-own-win32-app

After step 3 (modifying the vcxproj to include all the properties marked with "Island-support") I get the above mentioned build warning / compile error.

The part with the conflicting TRY macro I already encountered in our application when using the WinRT in general. We work around that by un- and redefining the macro around the includes in question (I really can't wait to use C++ modules). That has nothing to do with my current issue though.

**Reply by DarranRowe** [reply]
After looking through things, I would really suggest that you don't put the Xaml Islands functionality in the MFC executable.
That error that you mentioned is due to _AFXDLL becoming undefined. A superficial look at an MFC project shows that after you fix this error, it also then fails to link against WinMain. This is problematic because the default setup for MFC should have it trying to use the Unicode version and thus link against wWinMain. The entry point should also be provided by mfcs140u(d).lib. This makes me wonder if things will be subtly broken.
While it will end up requiring you to think about a DLL interface, I would seriously suggest you create a DLL especially for the Xaml Islands functionality and place it in there. You should be able to follow the steps provided with the add Xaml Content sample in that DLL. You should be able to reference C++/WinRT and the Windows App SDK in the MFC application too, but setting `<ApplicationType>Windows Store</ApplicationType>` is needed to enable the Xaml compiler, and this seems like it is getting in the way of MFC.
If you could investigate if the Windows App SDK itself can be referenced in your MFC application then it would help. Being able to use the Windows App SDK and the C++/WinRT projection on the DLL boundary would make your life much easier.
